ros2topic/ros2topic/verb/hz.py

```diff
@@ -0,0 +1,337 @@
+                     use_wtime=args.use_wtime)
+
+
+class ROSTopicHz(object):
+    """ROSTopicHz receives messages for a topic and computes frequency."""
+
+    def __init__(self, node, window_size, filter_expr=None, use_wtime=False):
+        import threading
+        from collections import defaultdict
+        self.lock = threading.Lock()
+        self.last_printed_tn = 0
+        self.msg_t0 = -1
+        self.msg_tn = 0
+        self.times = []
+        self._last_printed_tn = defaultdict(int)
+        self._msg_t0 = defaultdict(lambda: -1)
+        self._msg_tn = defaultdict(int)
+        self._times = defaultdict(list)
+        self.filter_expr = filter_expr
+        self.use_wtime = use_wtime
+
+        self.window_size = window_size
+
+        # Clock that has support for ROS time.
+        self._clock = node.get_clock()
+
+    def get_last_printed_tn(self, topic=None):
+        if topic is None:
+            return self.last_printed_tn
+        return self._last_printed_tn[topic]
+
+    def set_last_printed_tn(self, value, topic=None):
+        if topic is None:
+            self.last_printed_tn = value
+        self._last_printed_tn[topic] = value
+
+    def get_msg_t0(self, topic=None):
+        if topic is None:
+            return self.msg_t0
+        return self._msg_t0[topic]
+
+    def set_msg_t0(self, value, topic=None):
+        if topic is None:
+            self.msg_t0 = value
+        self._msg_t0[topic] = value
+
+    def get_msg_tn(self, topic=None):
+        if topic is None:
+            return self.msg_tn
+        return self._msg_tn[topic]
+
+    def set_msg_tn(self, value, topic=None):
+        if topic is None:
+            self.msg_tn = value
+        self._msg_tn[topic] = value
+
+    def get_times(self, topic=None):
+        if topic is None:
+            return self.times
+        return self._times[topic]
+
+    def set_times(self, value, topic=None):
+        if topic is None:
+            self.times = value
+        self._times[topic] = value
+
+    def callback_hz(self, m, topic=None):
+        """
+        Calculate interval time.
+
+        :param m: Message instance
+        :param topic: Topic name
+        """
+        # ignore messages that don't match filter
+        if self.filter_expr is not None and not self.filter_expr(m):
+            return
+        with self.lock:
+            # Uses ROS time as the default time source and Walltime only if requested
+            curr_rostime = self._clock.now() if not self.use_wtime else \
+                Clock(clock_type=ClockType.SYSTEM_TIME).now()
+
+            # time reset
+            if curr_rostime.nanoseconds == 0:
+                if len(self.get_times(topic=topic)) > 0:
+                    print('time has reset, resetting counters')
+                    self.set_times([], topic=topic)
+                return
+
+            curr = curr_rostime.nanoseconds
+            if self.get_msg_t0(topic=topic) < 0 or self.get_msg_t0(topic=topic) > curr:
```

Comment:
This would be slightly more efficient as:

```
msg_t0 = self.get_msg_t0(topic=topic)
if msg_t0 < 0 or msg_t0 > curr:
   ...
```

Reply:
thanks, fix in 0ba39d3